src/application/SessionController.cpp

```diff
@@ -0,0 +1,160 @@
+#include "SessionController.h"
+#include "ui/UIEventHandler.h"
+#include "System/ErrorLogger.h"
+
+#define DEFAULT_LOGGING_INTERVAL 100
+
+#define OUTPUT_FILENAME_ID      CONFIG_ID_OUTPUT_FILE_GLOBAL_ID
+#define VALUE_IDS               CONFIG_ID_VALUE_IDS
+#define LOGGING_INTERVAL_ID     CONFIG_ID_LOGGING_INTERVAL
+
+SessionController::SessionController(ApplicationContext& context, Adafruit_GFX& display,
+    SensorOptical& optical, SensorForce& force) : ControllerBase(context, display),
+        _optical(optical), _force(force) { }
+
+SessionController::~SessionController() { }
+
+void SessionController::init(InputManager& manager, StateInfo& info) {
+    // initialize inputs
+    ControllerBase::init(manager);
+    
+    // initialize output file
+    _initializeOutput(info);
+
+    // set logging interval
+    if (info.config.find(LOGGING_INTERVAL_ID) != info.config.end()) {
+        uint32_t interval = info.config[LOGGING_INTERVAL_ID].toInt();
+        dyno_assert(interval != 0);
+        _loggingInterval = interval;
+    } else {
+        dyno_log_str("SessionController: no logging interval defined -- setting to default " + String(DEFAULT_LOGGING_INTERVAL) + "ms");
+        _loggingInterval = DEFAULT_LOGGING_INTERVAL;
+    }
+
+    _info = info;
+}
+
+void SessionController::_logValues() {
```

Comment:
what do you think about adding an assert here to verify that the columnNum of dataLogger is 0? Just to make sure that we are still lined up properly

Reply:
I had the same thought, but if this does somehow go out of sync then it would spam the error log file because _logValues will be called many thousands of times while logging.

There are already tests for CsvFile::addEntry, so I trust that it won't fall out of line.  I have done some manual testing as well.  What I could do is make _outputCsv a private field so that it can only be controlled via protected SessionController methods.  It might also be worthwhile adding a unit test to make sure that everything is initialized properly.

Reply:
> What I could do is make _outputCsv a private field

I would be on board with this. But if we are going this direction, if we want to be completely safe I believe we would also have to restrict access to 
```
protected:
std::vector<std::function<String(void)>> _valueLoggers; // tandem with _valueIds
std::vector<ValueId> _valueIds;                         // tandem with _valueLoggers
```
to prevent messing with the sizes of the valueLoggers vector

Reply:
Okay, now that I think about it I think just making _outputCsv private is enough lol. I dont think it will be likely that people will be changing _valueLoggers after initialization.

Reply:
I still haven't built out the derived classes, but it might be worth making some changes to SessionController once those take shape.  In the meantime, _valueIds and _valueLoggers may be needed to update display values in derived classes.  I'll definitely revisit this after building out the calibration controller class